tools/golang/pkg/encryption/encryption.go

```diff
@@ -0,0 +1,295 @@
+/* Copyright 2019 The Subscribe with Google Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package encryption
+
+import (
+	"bytes"
+	"encoding/base64"
+	"encoding/json"
+	"errors"
+	"github.com/golang/protobuf/proto"
+	"github.com/google/tink/go/aead"
+	"github.com/google/tink/go/core/registry"
+	"github.com/google/tink/go/hybrid"
+	"github.com/google/tink/go/insecurecleartextkeyset"
+	"github.com/google/tink/go/keyset"
+	gcmpb "github.com/google/tink/proto/aes_gcm_go_proto"
+	tinkpb "github.com/google/tink/proto/tink_go_proto"
+	"golang.org/x/net/html"
+	"golang.org/x/net/html/atom"
+	"net/http"
+	"strings"
+	"unicode/utf8"
+)
+
+// Helper functions for the SwG Encryption Script.
+
+const aesGCMKeyURL string = "type.googleapis.com/google.crypto.tink.AesGcmKey"
+const aesGCMKeySize uint32 = 16
+
+// Public function to generate an encrypted HTML document given the original.
+func GenerateEncryptedDocument(htmlStr, accessRequirement string, pubKeys map[string]tinkpb.Keyset) (string, error) {
```

Comment:
How do you plan to version/evolve this over time? Backwards compatibility? Versioned encryption tag?

Reply:
I imagine that this will largely happen on the backend or in the AMP runtime. We can support multiple versions inside of the AMP Cache, and then handle each version specifically with the AMP extension: https://github.com/ampproject/amphtml/blob/master/extensions/amp-subscriptions/0.1/crypto-handler.js

Right now, we are supporting encrypted documents for up to five years after encryption. After that, we will redirect to origin.